### PR TITLE
Add Regular-Expression-Regex.html

### PR DESCRIPTION
**Pull request by pStyl3**
This new `.html` file was written in order to document, preserve and explain all kinds of Regular Expressions - aka "Regex" - we have previously used in some way or may use at a later date.

Any feedback is welcome, may it be of structural, logical or even grammatical nature.

Many thanks to @sibir-ine @MacSplody @Infernio @Arthmoor @3ventic and @BellCubeDev for helping me out and answering questions I had as well as providing examples. I've also found @kymckay and @egocarib's contributions from long ago very interesting (namely the BOSS -> LOOT masterlist conversion issue for SLE's masterlist).

### CODE REVIEW

File: docs/contributing/Regular-Expression-Regex.html [hunk 1]
```diff
@@ -0,0 +1,1140 @@
...
+title: Regular Expression - Regex
+---
+
+<p>The purpose of this page is to provide detailed information for using Regex while editing a masterlist, as well as searching specific terms within it. Regular expressions can formally be described as follows:</p>
+
+<blockquote cite="https://en.wikipedia.org/wiki/Regular_expression">
+A regular expression (shortened as regex or regexp) is a sequence of characters that specifies a search pattern in text.
+</blockquote>
+
+<h1>General</h1>
```
Comment (Ortham) [inline docs/contributing/Regular-Expression-Regex.html:12]: h1 renders too large:

image

Start with h3 as the largest heading in this page and work down from there.

File: docs/contributing/Regular-Expression-Regex.html [hunk 2]
```diff
@@ -0,0 +1,1140 @@
...
+      - link: 'https://www.nexusmods.com/skyrimspecialedition/mods/41713/'
+        name: 'DragonPriest Retexture'
+
+  - name: 'AcolyteMasks.esp'
+    url:
+      - link: 'https://www.nexusmods.com/skyrimspecialedition/mods/41713/'
+        name: 'DragonPriest Retexture'
+</code></pre>
+
+<p>This method however is not exactly efficient, as we have added the same link twice to the masterlist. With the help of a regex we can simplify the above expression.</p>
```
Comment (Ortham) [inline docs/contributing/Regular-Expression-Regex.html:35]: The mention of efficiency is ambiguous here: it's faster to find non-regex entries, but they take up more space. For a masterlist maintainer the main issue is maintainability, so maybe reword this to specifically mention how two entries are more of a maintenance burden than one?

File: docs/contributing/Regular-Expression-Regex.html [hunk 3]
```diff
@@ -0,0 +1,1140 @@
...
+  <li><code>\d</code> can be used exclusively for integer numbers 0 to 9</li>
+  <li><code>\D</code> can be used to identify (non-number) characters from the Latin Alphabet (except <code>Ä Æ Ö Ü</code>) and all the Special Characters</li>
+
+  <li><code>\w</code> can be used for the integer numbers 0 to 9, for any of the normal letters from the Latin Alphabet and for the underscore _</li>
+  <li><code>\W</code> can detect exclusively all the special characters except the underscore _ (and no numbers or Latin characters)</li>
+
+  <li><code>\s</code> detects whitespace</li>
+  <li><code>\S</code> can detect non-whitespace characters: Numbers, Latin characters (except additional ones) and the special characters</li>
+
+  <li>None of the expressions <code>\d \D \w \W \S</code> apply for the Greek and Japanese characters</li>
```
Comment (Ortham) [inline docs/contributing/Regular-Expression-Regex.html:94]: What characters are recognised varies by regex implementation used, so I think that these statements should be qualified to make it clear that they're talking about LOOT's use of regex in the name field of plugin objects in the masterlist.

Even then, LOOT uses three different regex implementations in different places, and this means:

* The search dialog uses Qt's regex implementation, which doesn't support unicode characters by default (but it can, I just didn't set the flag).
* The contains filter and masterlist entry matching uses C++'s standard regex, which doesn't support unicode.
* Condition evaluation uses the Rust regex library, which supports unicode by default.

File: docs/contributing/Regular-Expression-Regex.html [hunk 4]
```diff
@@ -0,0 +1,1140 @@
...
+<p>Japanese Hiragana:</p>
+<p><code>あ い う え お か き く け こ さ し す せ そ た ち つ て と な に ぬ ね の は ひ ふ へ ほ ま み む め も や ゆ よ ら り る れ ろ わ を ん</code></p>
+
+<p>Japanese Katakana:</p>
+<p><code>ア イ ウ エ オ カ キ ク ケ コ サ シ ス セ ソ タ チ ツ テ ト ナ ニ ヌ ネ ノ ハ ヒ フ ヘ ホ マ ミ ム メ モ ヤ ユ ヨ ラ リ ル レ ロ ワ ヲ ン</code></p>
+
+<p>Through testing the following rules could be derived:</p>
+
+<ul>
+  <li><code>\d</code> can be used exclusively for integer numbers 0 to 9</li>
```
Comment (Ortham) [inline docs/contributing/Regular-Expression-Regex.html:85]: You've used the term "integer numbers" throughout the document, but "digits" is the correct term.

File: docs/contributing/Regular-Expression-Regex.html [hunk 5]
```diff
@@ -0,0 +1,1140 @@
...
+
+<ul>
+  <li>The period <code>.</code> has been escaped through writing a backslash <code>\</code> in front of it.</li>
+  <li>The expression <code>\d{3,}</code> finds plugins that contain three or more integer numbers.</li>
+  <li>In addition to the regex this example also showcases a method to display blocks of code within a message for plugins.</li>
+  <li>The mentioned <code>Skyrim.ini</code> can be found in <code>C:\Users\USER\Documents\My Games\Skyrim</code></li>
+</ul>
+
+
+<h3>x to y times: <code>{x,y}</code></h3>
```
Comment (Ortham) [inline docs/contributing/Regular-Expression-Regex.html:279]: It would be good to explicitly state that this range is inclusive (though it's explained in the example summary).

File: docs/contributing/Regular-Expression-Regex.html [hunk 6]
```diff
@@ -0,0 +1,1140 @@
...
+  <li>As a first step the period <code>.</code> in front of <code>dll</code> has been escaped through writing a backslash <code>\</code> in front of it.</li>
+  <li>The remaining part of the regex <code>[^\.]+</code> consists out of three parts.</li>
+  <li><code>[^x]</code> will find a character that is <i>not x</i>, which is achieved by placing the circumflex <code>^</code> within square brackets <code>[ ]</code>. In this case <code>x</code> is an escaped period <code>\.</code> so <i>not a period</i> is what we want. After the square brackets comes the quantifier plus <code>+</code> which means <i>one or more times</i>.</li>
+  <li>Therefore the regular expression <code>[^\.]+\.dll</code> finds all SKSE plugins, which filesnames include one or more characters that are not a period, and end with <code>.dll</code></li>
+</ul>
+
+
+<h2>Lookarounds</h2>
+
+<h3>Negative lookahead: <code>(?! ...)</code></h3>
```
Comment (Ortham) [inline docs/contributing/Regular-Expression-Regex.html:572]: Like the Perl character classes at the start, support for lookaround varies by implementation and I think in LOOT negative lookahead is supported everywhere apart from inside conditions.

**Review by Ortham — CHANGES_REQUESTED** [review]

File: docs/contributing/Regular-Expression-Regex.html (hunk 1 above)
Comment (pStyl3) [reply docs/contributing/Regular-Expression-Regex.html]: I've adjusted the size of the different headings in https://github.com/loot/loot.github.io/pull/109/commits/383614f05462d908cddb5afdda9a4e9fd9149a69.

Specifically I replaced `<h3></h3>` with `<h5></h5>`, `<h2></h2>` with `<h4></h4>` and `<h1></h1>` with `<h3></h3>`.

File: docs/contributing/Regular-Expression-Regex.html [hunk 7]
```diff
@@ -0,0 +1,1140 @@
...
+  - name: 'AcolyteMasks.esl'
+    url:
+      - link: 'https://www.nexusmods.com/skyrimspecialedition/mods/41713/'
+        name: 'DragonPriest Retexture'
+
+  - name: 'AcolyteMasks.esp'
+    url:
+      - link: 'https://www.nexusmods.com/skyrimspecialedition/mods/41713/'
+        name: 'DragonPriest Retexture'
+</code></pre>
```
Comment (BellCubeDev) [inline docs/contributing/Regular-Expression-Regex.html:33]: For prettier code blocks, you can use Jekyll's `{% highlight %}` tag. To actually render the highlighting, though, you would need to add the relevant classes to the site-wide CSS.

For this instance, the code block would look like this:
```suggestion
<pre>{% highlight yaml %}
  - name: 'AcolyteMasks.esl'
    url:
      - link: 'https://www.nexusmods.com/skyrimspecialedition/mods/41713/'
        name: 'DragonPriest Retexture'

  - name: 'AcolyteMasks.esp'
    url:
      - link: 'https://www.nexusmods.com/skyrimspecialedition/mods/41713/'
        name: 'DragonPriest Retexture'
{% endhighlight %}</pre>
```
(don't worry - Jekyll adds the `<code>` tag back)

The classes I've found:
* `pi` - Characters with special meaning such as `:` , `>`, `|`, and `-` outside of a string
* `s1` - Single-Quotes (`'`)
* `s2` - Double-Quotes (`"`)
* `na` - Keys
* `s` - String or Number values
* `nv` - Appears to be whitespace within a quoted string
* `no` - `null` and similar values

File: docs/contributing/Regular-Expression-Regex.html (hunk 4 above)
Comment (pStyl3) [reply docs/contributing/Regular-Expression-Regex.html]: I've pushed https://github.com/loot/loot.github.io/pull/109/commits/ee5402825567bfd978adecebf46adc18231984d1 (and https://github.com/loot/loot.github.io/pull/109/commits/02a9bea3a3e6743ba4bbb9c1ce874e0967eb0df3) for this. I think it should be good now, but if you think that it can be improved further, don't hesitate to tell me.

Comment (pStyl3) [reply docs/contributing/Regular-Expression-Regex.html]: Additional changes: https://github.com/loot/loot.github.io/pull/109/commits/cf56f0a40cc4ae630160fdea0bd31b5559e0cf17

File: docs/contributing/Regular-Expression-Regex.html (hunk 2 above)
Comment (pStyl3) [reply docs/contributing/Regular-Expression-Regex.html:35]: Good point, I haven't thought about that. How about this:

> <p>This writing-style is a valid method to add metadata to the masterlists and it also means that searching for the plugin entry (e.g. <code>AcolyteMasks.esl</code>) is fast and easy, as it is exactly the same as the actual plugin name.
> 
> <p>The downside to it is however, that these non-regex entries take up more space than their regex counterparts, as we have added the same link twice to the masterlist. In order to keep things maintainable, we often choose to make use of regular expressions such as the following one:</p>

I think that the word "burden" has a pretty strong negative connotation to it, as such I've tried to avoid using it.

File: docs/contributing/Regular-Expression-Regex.html (hunk 5 above)
Comment (pStyl3) [reply docs/contributing/Regular-Expression-Regex.html]: Done in https://github.com/loot/loot.github.io/pull/109/commits/93dc3c9a604aa8a1f77bc25aadf87d808df1b1d9.

File: docs/contributing/Regular-Expression-Regex.html (hunk 7 above)
Comment (pStyl3) [reply docs/contributing/Regular-Expression-Regex.html:33]: yaml highlighting would be great indeed, and our [Masterlist Editing](https://loot.github.io/docs/contributing/Masterlist-Editing.html) file could profit from that as well. However, I'm not sure how to go about this, specifically how I would integrate Jekyll support in order to use these `{% highlight yaml %}{% endhighlight %}` tags.

There's the `\css\style.css` file .. but I don't have experience with that.

Comment (BellCubeDev) [reply docs/contributing/Regular-Expression-Regex.html:33]: Jekyll is the site builder that you're already using (it's built into GitHub Pages)—that's what the `layout: page` stuff at the top of the document is all about. The tag, for our purposes, is plug-and-play. As for the CSS, I won't be around for about a week, so I unfortunately can't be the one to write it. You could probably get someone—possibly on the Discord server—to write it though!

File: docs/contributing/Regular-Expression-Regex.html (hunk 2 above)
Comment (Ortham) [reply docs/contributing/Regular-Expression-Regex.html:35]: Yeah, that's a better description of the relevant pros and cons.

File: docs/contributing/Regular-Expression-Regex.html (hunk 7 above)
Comment (Ortham) [reply docs/contributing/Regular-Expression-Regex.html:33]: Personally, I find that the easiest way to run Jekyll on Windows is using Docker, for which I use the following command:

```
docker run --rm -v ${PWD}:/srv/jekyll -v jekyll_bundle:/usr/local/bundle -p 4000:4000 -it jekyll/jekyll:3.8.6 jekyll serve --no-watch
```

File: docs/contributing/Regular-Expression-Regex.html (hunk 2 above)
Comment (pStyl3) [reply docs/contributing/Regular-Expression-Regex.html]: Done in https://github.com/loot/loot.github.io/pull/109/commits/d7236457470136fd700ad24130ca86362d5e21ce,

**Comment by pStyl3** [comment]
One existing example has been further improved in https://github.com/loot/loot.github.io/pull/109/commits/6ccdd7fbfab34975013beb8aca02bb2f40be81dd and I've also added one more example for the Find & Replace section in https://github.com/loot/loot.github.io/pull/109/commits/567d7999eb90da02faa1100c3daa18f52948849b.

**Review by Ortham — APPROVED** [review]

File: docs/contributing/Regular-Expression-Regex.html (hunk 6 above)
Comment (pStyl3) [reply docs/contributing/Regular-Expression-Regex.html]: I've added a notice in https://github.com/loot/loot.github.io/pull/109/commits/0e54156b2c3af237cf3db7c585e1bdef3d531e19.

File: docs/contributing/Regular-Expression-Regex.html (hunk 7 above)
Comment (pStyl3) [reply docs/contributing/Regular-Expression-Regex.html:33]: I've heard of Docker before, but haven't used that one before either. I'm open to update this file at a later date to implement the `yaml` highlighting, however I would need help for that as I currently don't understand all the implications on working with CSS/Jekyll/Docker. If you've got the time @BellCubeDev and are interested, please ping me once you are around. 😄

**Comment by pStyl3** [comment]
I've marked the PR as ready for review, as I think that the file should be good now.

File: docs/contributing/Regular-Expression-Regex.html [hunk 8]
```diff
@@ -53,6 +53,40 @@ <h4>Advantages of Regex</h4>
...
+        condition: 'many("Example\.es(m|p)")'
+</code></pre>
+
+<p>The regex in the <code>name</code> field uses <code>C++</code>'s standard regex, which doesn't support unicode.</p>
+
+<p>The regex in the <code>condition</code> field uses the <code>Rust</code> regex library, which supports unicode by default.</p>
+
+<img src="../../images/Regex_Search_Cards.png" alt="Regex_Search_Cards" width="602" height="409">
+
+<p>The search dialog uses <code>Qt</code>'s regex implementation, which doesn't support unicode characters by default (though support could be enabled).</p>
```
[the remark this answers is not in the document]
Comment (pStyl3) [reply docs/contributing/Regular-Expression-Regex.html]: Done in https://github.com/loot/loot.github.io/pull/109/commits/9a9cdf38fef21826797ac9e35dd3e86dda1c5f35.
